### PR TITLE
Many small deriving cleanups

### PR DESCRIPTION
**Pull request by nnethercote**
These commits remove lots of little unnecessary things, and clarifies other things.

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by nnethercote** [comment]
Best reviewed one commit at a time.

I have a bunch of improvement to the code generation that depend on these changes.

**Comment by Mark-Simulacrum** [comment]
A couple nits, r=me with them resolved. (Unless you think resolution merits another review).

File: compiler/rustc_builtin_macros/src/deriving/clone.rs [hunk 1]
```diff
@@ -15,20 +15,19 @@ pub fn expand_deriving_clone(
     item: &Annotatable,
     push: &mut dyn FnMut(Annotatable),
 ) {
-    // check if we can use a short form
+    // The shallow form is `fn clone(&self) -> Self { *self }`, possibly with
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/deriving/clone.rs:18]: "Shallow" comes from the existing `cs_clone_shallow` function. But you're right that it's a confusing name, so I will change all the occurrences to "simple".

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs [hunk 2]
```diff
@@ -224,10 +224,8 @@ pub struct MethodDef<'a> {
     /// List of generics, e.g., `R: rand::Rng`
     pub generics: Bounds,
 
-    /// Whether there is a self argument (outer Option) i.e., whether
-    /// this is a static function, and whether it is a pointer (inner
-    /// Option)
-    pub explicit_self: Option<Option<PtrTy>>,
+    /// Is there is a `&self` argument? If not, it is a static function.
+    pub explicit_self: bool,
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:228]: "Explicit self" is a phrase that appears in several other places in the compiler. There's even a type `ast::ExplicitSelf`. Plus it's a pre-existing name, so I'll leave it unchanged.

**Comment by nnethercote** [comment]
Thank you for the fast review.

File: compiler/rustc_builtin_macros/src/deriving/generic/mod.rs [hunk 3]
```diff
@@ -1717,22 +1717,21 @@ where
...
+                    let (first, rest) = all_fields.split_first().unwrap();
+                    let args = (first.span, first.self_.clone(), &first.other[..]);
+                    (b(cx, Some(args)), rest)
                 }
                 (false, false) => {
-                    let idx = all_fields.len() - 1;
-                    let field = &all_fields[idx];
-                    let args = (field.span, field.self_.clone(), &field.other[..]);
-                    (b(cx, Some(args)), &all_fields[..idx])
+                    let (last, rest) = all_fields.split_last().unwrap();
```
[the remark this answers is not in the document]
Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/deriving/generic/mod.rs:1669]: I'll leave it unchanged.

**Pull request merged**